File: objetismo.py
#!/usr/bin/python3
# -*- coding: utf-8 -*- 

import pygame
import datetime
import sys
import struct
import math
import ConfigParser
import argparse


surfaces = {}
objects = {}
colors = {}
cache = {}


def getAverageColor(surface):
    r, g, b = 0, 0, 0
    count = 0
    for x in xrange(surface.get_width()):
        for y in xrange(surface.get_height()):
            cR, cG, cB, alpha = surface.get_at((x, y))
            if cR+cG+cB == 0: continue
         
            r += cR
            g += cG
            b += cB
            count += 1
    if (count == 0): 
        return (0,0,0,1)
    return (r/count), (g/count), (b/count), count+1


def saveSurface(surface, filename):
    pygame.image.save(surface, filename)

def getColor(color):
    r = color[0]
    g = color[1]
    b = color[2]
    return r + g*256 + b*65536 

def loadObjectGraphic(objectId):
    if objectId == None: return None
    grh = int(objects[objectId]['graphic'])
    grhData = surfaces[grh]
    objectSurface = pygame.Surface((32, 32))
    objectSurface.blit(
                       pygame.image.load(graphics + str(grhData['file']) + '.bmp'),
                       (0, 0, 32, 32), 
                       (grhData['sX'], grhData['sY'], grhData['sX'] + 32, grhData['sY'] + 32)
                       )
    return objectSurface

def RGB_to_XYZ(rgb):
    
    r = rgb[0] / 255.0
    g = rgb[1] / 255.0
    b = rgb[2] / 255.0
    
    if r > 0.04045:
        r = math.pow((r + 0.055) / 1.055, 2.4)
    else:
        r = r / 12.92
    if g > 0.04045:
        g = math.pow((g + 0.055) / 1.055, 2.4)
    else:
        g = g / 12.92
    if b > 0.04045:
        b = math.pow((b) / 1.055, 2.4)
    else:
        b = b / 12.92
      
    return (r, g, b)

def XYZ_to_Lab(xyz):

    x = xyz[0]/  0.95047
    y = xyz[1] / 1.00000
    z = xyz[2] / 1.08883
    
    CIE_E = 0.008856 #216.0/24389.0
    
    if x > CIE_E:
        x = math.pow(x, (1.0 / 3.0))
    else:
        x = (7.787 * x) + (16.0 / 116.0)     

    if y > CIE_E:
        y = math.pow(y, (1.0 / 3.0))
    else:
        y = (7.787 * y) + (16.0 / 116.0)
   
    if z > CIE_E:
        z = math.pow(z, (1.0 / 3.0))
    else:
        z = (7.787 * z) + (16.0 / 116.0)
      
    l = (116.0 * y) - 16.0
    a = 500.0 * (x - y)
    b = 200.0 * (y - z)

    return (l,a,b)

def cie1976(color1, color2):
    r1,g1,b1,alpha1 = color1
    r2,g2,b2,alpha2 = color2
    lab1 = XYZ_to_Lab(RGB_to_XYZ((r1,g1,b1)))
    lab2 = XYZ_to_Lab(RGB_to_XYZ((r2,g2,b2)))
    
    return math.sqrt((lab1[0]-lab2[0])**2 + (lab1[1]-lab2[1])**2 + (lab1[2]-lab2[2])**2)

def cie1994(color1, color2):
    r1,g1,b1,alpha1 = color1
    r2,g2,b2,alpha2 = color2

    l1,a1,b1 = XYZ_to_Lab(RGB_to_XYZ((r1,g1,b1)))
    l2,a2,b2 = XYZ_to_Lab(RGB_to_XYZ((r2,g2,b2)))

    dL = l1 - l2
    dA = a1 - a2
    dB = b1 - b2
    c1 = math.sqrt(a1**2 + b1**2)
    c2 = math.sqrt(a2**2 + b2**2)
    dC = c1 - c2
    dH2 = dA**2 + dB**2 - dC**2
    if dH2 > 0.0:
        dH = math.sqrt(dH2)
    else:
        dH = 0.0
    sL = 1.0
    k1 = 0.045 #0.048
    k2 = 0.015 #0.014
    sC = 1.0 + k1 * c1
    sH = 1.0 + k2 * c1
    
    kL = 1.0
    kC = 1.0
    kH = 1.0
    
    e = (dL / (kL * sL))**2 + (dC / (kC * sC))**2 + (dH / (kH * sH))**2
    
    return math.sqrt(e)

def cie2000(color1,color2):
    r1,g1,b1,alpha1 = color1
    r2,g2,b2,alpha2 = color2

    l1,a1,b1 = XYZ_to_Lab(RGB_to_XYZ((r1,g1,b1)))
    l2,a2,b2 = XYZ_to_Lab(RGB_to_XYZ((r2,g2,b2)))
    
    L = 1.0 # 2
    C = 1.0 
    
    c1 = math.sqrt(a1**2 + b1**2)
    c2 = math.sqrt(a2**2 + b2**2)
    if l1 < 16.0:
        sL = 0.511 
    else:
        sL = (0.040975 * l1) / (1.0 + 0.01765 * l1)
    sC = (0.0638 * c1) / (1.0 + 0.0131 * c1) + 0.638;
    if c1 < 0.000001:
        h1 = 0.0
    else:
        h1 = (math.atan2(b1, a1) * 180.0) / math.pi
        
    while (h1 < 0.0):
        h1 += 360.0
    while (h1 >= 360.0):
        h1 -= 360.0;
        
    if (h1 >= 164.0) and (h1 <= 345.0):
        t = (0.56 + abs(0.2 * math.cos((math.pi * (h1 + 168.0)) / 180.0)))
    else: 
        t = (0.36 + abs(0.4 * math.cos((math.pi * (h1 + 35.0)) / 180.0)))
        
    c4 = c1**4
    
    f = math.sqrt(c4 / (c4 + 1900.0));
    sH = sC * (f * t + 1.0 - f)
    dL = l1 - l2;
    dC = c1 - c2;
    dA = a1 - a2
    dB = b1 - b2
    dH2 = dA**2 + dB**2 - dC**2
    v1 = dL / (L * sL);
    v2 = dC / (C * sC);
    v3 = sH;
    
    if (L == 1.0):
        return math.sqrt(v1 * v1 + v2 * v2 + (dH2 / (v3 * v3)));
    else:
        return math.sqrt(v1 * v1 + v2 * v2 + (dH2 / (v3 * v3)));

def colorDistance(color1, color2):

    return algorithm(color1, color2)



def getObjectForColor(color):
    colorValue = getColor(color)
    try:
        r = cache[colorValue]
        # print(color)
        return r
    except:
        d = 9223372036854775807 
        ret = None
        for c in colors:
            cD = colorDistance(color, colors[c])
            
            # Same color
            if cD == 0:
                ret = c
                break;

            if cD < d:
                d = cD
                ret = c

    cache[colorValue] = loadObjectGraphic(ret)
    return cache[colorValue] 
        
        
def loadGraphics(tolerance):
    for object in objects:
        if object == None or objects[object] == None: continue
        
        grh = int(objects[object]['graphic'])

        if (grh == None or grh <= 0 or surfaces[grh] == None): continue
        
        s = loadObjectGraphic(object)
        color = getAverageColor(s)
    
        # Descarto los objetos que tienen mucho negro (ejemplo: llaves)
        if color[3] > 32*32*tolerance/100.0: 
            colors[object] = color

def loadObjects(dats):
    config = ConfigParser.ConfigParser()
    config.read(dats)
    maxObjects = int(config.get('INIT', 'NumOBJs'))
    for i in xrange(1, maxObjects):
        sectionName = 'OBJ' + str(i)
        if not config.has_section(sectionName):
            objects[i] = None
        else:
            objects[i] = {
                'name': config.get(sectionName, 'Name'),
                'graphic': config.get(sectionName, 'GrhIndex') 
            }

def loadGraphicsData(indexesFile):
    
    file = open(indexesFile, 'r')
    
    struct.unpack('i', file.read(4))[0]
    
    grhCount = struct.unpack('i', file.read(4))[0]
    grh = 0
    data = None
    while grhCount-1 != grh:
        grh = struct.unpack('i', file.read(4))[0]
        if grh != 0:
            framesAmount = struct.unpack('h', file.read(2))[0]
            if framesAmount <= 0:  return

            # animation
            if framesAmount > 1:
                for i in range(framesAmount):
                    frame = struct.unpack('i', file.read(4))[0]
                    if frame <= 0 or frame > grhCount: return;                        
                
                speed = struct.unpack('i', file.read(4))
                if (speed <= 0): return
                
                data = None
            else:
                fileNumber = struct.unpack('i', file.read(4))[0]
                if fileNumber <= 0: return
                
                sX = struct.unpack('h', file.read(2))[0]
                if sX < 0: return
                
                sY = struct.unpack('h', file.read(2))[0]
                if sY < 0: return
                
                tileWidth = struct.unpack('h', file.read(2))[0]
                if tileWidth <= 0: return
                
                tileHeight = struct.unpack('h', file.read(2))[0]
                if tileHeight <= 0: return
                
                # exclude non tile-size graphics.
                if tileHeight == 32 and tileWidth == 32:
                    data = {
                        'file': fileNumber,
                        'sX': sX,
                        'sY': sY,
                        'width': tileWidth,
                        'height': tileHeight
                    }
                else:
                    data = None

            surfaces[grh] = data


parser = argparse.ArgumentParser(description='Objetismo')
parser.add_argument('--image', '-i', required=True)
parser.add_argument('--algo', '-a', default=3, type=int, help='Opcional. Hay 6 algoritmos: CIE 1976 (1), CIE 1994 (2)')
parser.add_argument('--tolerance', '-t', default=30)
parser.add_argument('--graphics', '-g', default='resources/graficos/')
parser.add_argument('--dats', '-d', default='resources/obj.dat')
parser.add_argument('--index', '-n', default='resources/Graficos3.ind')
parser.add_argument('--output', '-o', default='output/')

# 
args = parser.parse_args()
filename = args.image
tolerance = args.tolerance
graphics = args.graphics
dats = args.dats
index = args.index
if args.algo == 1:
    algorithm = cie1976
elif args.algo == 2:
    algorithm = cie1994
elif args.algo == 3:
    algorithm = cie2000


pygame.init()

loadGraphicsData(index)
loadObjects(dats)
loadGraphics(tolerance)

image = pygame.image.load(filename)

w,h = image.get_size()

s = pygame.Surface((w*32, h*32))

for y in xrange(h):
    print str(y+1) + " / " + str(h)
    for x in xrange(w):
        obj = getObjectForColor(image.get_at((x,y)))
        s.blit(obj, (x*32, y*32))
        
    
now = datetime.datetime.now()
outputFile = 'output/' + now.strftime("%Y-%m-%d %H:%M:%S") + '.png'
print 'Guardando imagen en ' + outputFile
saveSurface(s, outputFile)

File: objetismo.cpp
#include <math.h>
#include <fstream>
#include <iostream>
#include <string>
#include <ctime>
#include <assert.h>

#include "png.h"

#define CIE_8 0.008856 // 216.0/24389.0

using namespace std;

struct RGB {
    int red;
    int green;
    int blue;

    // soy negro.
    RGB() : red(-1337), green(0), blue(0) {}

    RGB(int r, int g, int b) : red(r), green(g), blue(b) {
        if (longColor() == 16777215) red = green = blue = 0; 
    }

    RGB(const RGB& c) : red(c.red), green(c.green), blue(c.blue) {
        if (longColor() == 16777215) red = green = blue = 0;
    }

    // soy negro.
    bool valid() { return red >= 0; }

    int longColor() {
        return red + green*256 + blue*65536;
    }
};

struct XYZ {
    double x;
    double y;
    double z;

    XYZ(double x, double y, double z) : x(x), y(y), z(z) {}

};

struct LAB {
    double l;
    double a;
    double b;

    LAB(double l, double a, double b) : l(l), a(a), b(b) {}

};

int cache[256 + 256*256 + 256*65536] = {0};
RGB objects[2000];
int files[2000];


double xyzTransformation(double v) {
    if (v > 0.04045) {
        v = pow((v + 0.055) / 1.055, 2.4);
    } else {
        v = v / 12.92;
    } 
    return v;
}

XYZ RGB_to_XYZ(const RGB& c) {
    XYZ xyz(
        xyzTransformation(c.red / 255.0),
        xyzTransformation(c.green / 255.0),
        xyzTransformation(c.blue / 255.0)
    );

    return xyz;
}

double labTransformation(double v) {
    if (v > 0.008856) {
        // v = pow(v, 1/3.0);
        v = pow(v, 1/3.0);
    } else {
        v = (7.787 * v) + (16.0 / 116.0);
    }
    return v;
}

LAB XYZ_to_Lab(const XYZ& c) {
    double x = c.x / 0.95047;
    double y = c.y / 1.0;
    double z = c.z / 1.08883;

    x = labTransformation(x);
    y = labTransformation(y);
    z = labTransformation(z);

    LAB lab((116.0 * y) - 16, 500.0 * (x - y), 200.0 * (y - z));

    return lab;
}

int cie1976(const RGB& color1, const RGB& color2) {
    LAB lab1 = XYZ_to_Lab(RGB_to_XYZ(color1));
    LAB lab2 = XYZ_to_Lab(RGB_to_XYZ(color2));

    double d = sqrt( pow(lab1.l - lab2.l, 2) + pow(lab1.a - lab2.a, 2) + pow(lab1.b - lab2.b, 2) );

    return (int) d;
}

int cie1994(const RGB& color1, const RGB& color2) {

    LAB lab1 = XYZ_to_Lab(RGB_to_XYZ(color1));
    LAB lab2 = XYZ_to_Lab(RGB_to_XYZ(color2));

    double dL = lab1.l - lab2.l;
    double dA = lab1.a - lab2.a;
    double dB = lab1.b - lab2.b;
    double c1 = sqrt(pow(lab1.a, 2) + pow(lab1.b, 2));
    double c2 = sqrt(pow(lab2.a, 2) + pow(lab2.b, 2));
    double dC = c1 - c2;
    double dH2 = pow(dA, 2) + pow(dB, 2) - pow(dC, 2);
    double dH = 0;
    if (dH2 > 0.0) {
        // dH = math.sqrt(dA**2 + dB**2 - dC**2)
        dH = sqrt(dH2);
    }
    double sL = 1.0;
    double k1 = 0.045; //0.048
    double k2 = 0.015; // #0.014
    double sC = 1.0 + k1 * c1;
    double sH = 1.0 + k2 * c1;
    
    double kL = 1.0;
    double kC = 1.0;
    double kH = 1.0;
    
    // e = pow(dL / (kL * sL), 2) + pow(dC / (kC * sC), 2) + pow(dH / (kH * sH), 2)
    return (int) sqrt(pow(dL / (kL * sL), 2) + pow(dC / (kC * sC), 2) + pow(dH / (kH * sH), 2));
}

int colorDistance(const RGB& c1, const RGB& c2) {
    // return cie1994(c1, c2);
    return cie1976(c1, c2);
}

int getObjectForColor(RGB c) {
    int colorValue = c.longColor();
    int r = cache[colorValue];
    
    if (r != 0) {
        return r;
    }

    int i, cD;
    int d = 16777215;
    int best = -1;

    for (i = 0; i < 2000; i++) {
        if (!objects[i].valid()) continue;
        cD = colorDistance(c, objects[i]);
        // Same color!
        if (cD == 0) {
            best = i;
            break;
        }
        if (cD < d) {
            d = cD;
            best = i;
        }
    }
    if (best > 2000) {  
        std::cout << "assert! " << best << " for " << colorValue << std::endl;
        assert(false);
    }
    cache[colorValue] = best;
    return cache[colorValue];
}


// Get current date/time, format is YYYY-MM-DD.HH:mm:ss
const std::string currentDateTime() {
    time_t     now = time(0);
    struct tm  tstruct;
    char       buf[80];
    tstruct = *localtime(&now);
    // Visit http://en.cppreference.com/w/cpp/chrono/c/strftime
    // for more information about date/time format
    strftime(buf, sizeof(buf), "%Y-%m-%d.%X", &tstruct);
    return buf;
}

int main(int argc, const char *argv[]) {
    if (argc < 2) {
        std::cout << "Usage ./objetismo image.png" << std::endl;
        return 1;
    }


    ::std::ifstream in("resources/colors", ::std::ios::binary);

    while (!in.eof()) {
        int id1 = in.get();
        if (id1 == -1) break;
        int id2 = in.get();
        int id = id1 + (id2 * 256);
        int fi1 = in.get();
        int fi2 = in.get();
        int fi = fi1 + (fi2 * 256);

        int r = in.get();
        int g = in.get();
        int b = in.get();
        objects[id] = RGB(r, g, b);
        files[id] = fi;
    }

    PNG img;
    img.read(argv[1]);

    short w = img.width();
    short h = img.height();
    int x,y,counter = 0;
    short dst_matrix [w][h];
    int i = 0;
    int max = w*h;

    PNG* surfaces[2000];
    for (i = 0; i < 2000; i++) {
        surfaces[i] = NULL;
    }

    for (x = 0; x < w; x++) {
        for (y = 0; y < h; y++) {
            dst_matrix[x][y] = -1;
        }
    }

#pragma omp parallel for
    for (x = 0; x < w; x++) {
        std::cout << "d: " << counter << endl;
        for (y = 0; y < h; y++) {
            png_bytep p = img.getPixel(x, y);
            RGB c(p[0],p[1],p[2]);
            dst_matrix[x][y] = getObjectForColor(c);
            if (dst_matrix[x][y] > 2000) {
                std::cout << "assert! " << x << "," << y << std::endl;
                assert(false);
            }
        }
        counter++;
    }
    std::cout << "checking" << std::endl;

    for (x = 0; x < w; x++) {
        for (y = 0; y < h; y++) {
            if (dst_matrix[x][y] > 2000) {
                std::cout << x << "," << y << std::endl;
            }
        }
    }

    std::string res = "resources/graficos/";
    std::string output = "output/";
    std::string ext = ".png";
    PNG* sp;
    std::cout << "sprites" << std::endl;
    counter = 0;
    for (x = 0; x < w; x++) {
        // std::cout << "s: " << counter << endl;
        for (y = 0; y < h; y++) {
            short objId = dst_matrix[x][y];
            if (objId > 0 && surfaces[objId] == NULL) {
                sp = new PNG();
                std::string str = res + std::to_string(files[objId]) + ext;
                const char* filename = str.c_str();
                std::cout << "sprite " << objId << "..." << "\r";
                sp->read(filename);
                surfaces[objId] = sp;
                std::cout << "sprite " << objId << "... " << "Done!" << endl;                
            }
        }
        // counter++;
    }

    PNG image;
    image.create(w*32, h*32);
    std::cout << std::endl;
    for (x = 0; x < w; x++) {
        for (y = 0; y < h; y++) { 
           short objId = dst_matrix[x][y];
            if (objId < 0 || objId > 2000) {
                std::cout << "assert! " << objId << " en " << x << "," << y << std::endl;
                assert(false);
            }
            image.paste(*surfaces[objId], x*32, y*32);
        }
    }
    // const char* f = currentDateTime().c_str();
    // std::cout << f << std::endl; 
    std::cout << std::endl;
    image.write("bbbbbbbb.png");
    PNG* ap; 
    for (i = 0; i < 2000; i++) {
        if (surfaces[i] != NULL) delete surfaces[i];
    }

    return 0;
}
